File: logicposintegration/templates/pages/projects.js
frappe.ready(function () {
	var STATUS_COLORS = {
		Open: "orange",
		Working: "blue",
		"Pending Review": "yellow",
		Overdue: "red",
		Completed: "green",
		Cancelled: "grey",
		Template: "purple",
	};

	var PRIORITY_COLORS = {
		Low: "grey",
		Medium: "blue",
		High: "orange",
		Urgent: "red",
	};

	var gantt_tasks = [];
	var gantt_task_map = {};
	var gantt_i18n = {};

	try {
		gantt_i18n = JSON.parse(document.getElementById("project-gantt-i18n").textContent || "{}");
	} catch (e) {
		gantt_i18n = {};
	}

	function gantt_label(key, fallback) {
		return gantt_i18n[key] || fallback || key;
	}

	function gantt_status_label(status) {
		return (gantt_i18n.status_labels && gantt_i18n.status_labels[status]) || status;
	}

	function gantt_priority_label(priority) {
		return (gantt_i18n.priority_labels && gantt_i18n.priority_labels[priority]) || priority;
	}

	try {
		gantt_tasks = JSON.parse(document.getElementById("project-gantt-data").textContent || "[]");
		gantt_tasks.forEach(function (task) {
			gantt_task_map[String(task.id)] = task;
		});
	} catch (e) {
		gantt_tasks = [];
	}

	var project_gantt = null;
	var views = {
		list: "#tasks-list-view",
		gantt: "#tasks-gantt-view",
		kanban: "#tasks-kanban-view",
	};

	function show_view(view) {
		Object.keys(views).forEach(function (key) {
			$(views[key]).hide();
		});

		if (view === "gantt") {
			$(views.gantt).show();
			render_gantt();
		} else if (view === "kanban") {
			$(views.kanban).show();
		} else {
			$(views.list).show();
		}
	}

	$(".project-tasks-view-tab").on("click", function (e) {
		e.preventDefault();
		var view = $(this).data("view");
		$(".project-tasks-view-tab").removeClass("active");
		$(this).addClass("active");
		show_view(view);
	});

	$(".project-kanban-card").on("click", function () {
		var task = $(this).data("task");
		if (task) {
			window.location.href = "/tasks/" + encodeURIComponent(task);
		}
	});

	$(".portal-gantt-mode-btn").on("click", function () {
		var mode = $(this).data("mode");
		$(".portal-gantt-mode-btn").removeClass("active");
		$(this).addClass("active");
		if (project_gantt) {
			project_gantt.change_view_mode(mode);
		}
	});

	function get_task_meta(task) {
		var mapped = gantt_task_map[String(task.id)] || {};
		return {
			status: mapped.status || task.status || "",
			priority: mapped.priority || task.priority || "",
			progress: mapped.progress != null ? mapped.progress : task.progress,
			assignees: mapped.assignees || task.assignees || [],
		};
	}

	function build_status_pill(status) {
		if (!status) {
			return '<span class="text-muted">—</span>';
		}
		var color_key = STATUS_COLORS[status] || "grey";
		return (
			'<span class="portal-status-pill portal-status-' +
			color_key +
			'">' +
			frappe.utils.escape_html(gantt_status_label(status)) +
			"</span>"
		);
	}

	function build_priority_pill(priority) {
		if (!priority) {
			return '<span class="text-muted">—</span>';
		}
		var color_key = PRIORITY_COLORS[priority] || "grey";
		return (
			'<span class="portal-priority-pill portal-priority-' +
			color_key +
			'">' +
			frappe.utils.escape_html(gantt_priority_label(priority)) +
			"</span>"
		);
	}

	function build_assignee_avatars(assignees) {
		if (!assignees || !assignees.length) {
			return '<span class="text-muted">—</span>';
		}

		return (
			'<div class="portal-gantt-popup-assignees">' +
			assignees
				.map(function (assignee) {
					var name = assignee.full_name || assignee;
					var title = frappe.utils.escape_html(name);
					if (typeof assignee === "string") {
						return (
							'<span class="avatar avatar-small portal-avatar" title="' +
							title +
							'"><div class="standard-image">' +
							frappe.utils.escape_html(name.slice(0, 2).toUpperCase()) +
							"</div></span>"
						);
					}
					if (assignee.user_image) {
						return (
							'<span class="avatar avatar-small portal-avatar" title="' +
							title +
							'"><img src="' +
							frappe.utils.escape_html(assignee.user_image) +
							'" alt=""></span>'
						);
					}
					return (
						'<span class="avatar avatar-small portal-avatar" title="' +
						title +
						'"><div class="standard-image">' +
						frappe.utils.escape_html(assignee.abbr || name.slice(0, 2).toUpperCase()) +
						"</div></span>"
					);
				})
				.join("") +
			"</div>"
		);
	}

	function style_gantt_popup() {
		var popup = document.querySelector("#tasks-gantt-view .popup-wrapper");
		if (!popup) {
			return;
		}

		popup.style.setProperty("background", "#fff", "important");
		popup.style.setProperty("color", "#0f172a", "important");
		popup.style.setProperty("border", "1px solid #e2e8f0", "important");
		popup.style.setProperty("border-radius", "12px", "important");
		popup.style.setProperty("box-shadow", "0 12px 32px rgba(15, 23, 42, 0.12)", "important");
		popup.style.setProperty("padding", "0", "important");

		var pointer = popup.querySelector(".pointer");
		if (pointer) {
			pointer.style.display = "none";
		}
	}

	function build_gantt_popup(task) {
		var meta = get_task_meta(task);
		var progress = meta.progress != null ? meta.progress + "%" : "—";
		var task_url = "/tasks/" + encodeURIComponent(task.id);

		return (
			'<div class="portal-gantt-popup">' +
			'<div class="portal-gantt-popup-header">' +
			'<div class="portal-gantt-popup-title">' +
			frappe.utils.escape_html(task.name) +
			"</div>" +
			'<div class="portal-gantt-popup-dates">' +
			frappe.utils.escape_html(format_gantt_date(task._start)) +
			" → " +
			frappe.utils.escape_html(format_gantt_date(task._end)) +
			"</div></div>" +
			'<div class="portal-gantt-popup-body">' +
			'<div class="portal-gantt-popup-row"><span class="portal-gantt-popup-label">' +
			gantt_label("status", "Status") +
			'</span><span class="portal-gantt-popup-value">' +
			build_status_pill(meta.status) +
			"</span></div>" +
			'<div class="portal-gantt-popup-row"><span class="portal-gantt-popup-label">' +
			gantt_label("priority", "Priority") +
			'</span><span class="portal-gantt-popup-value">' +
			build_priority_pill(meta.priority) +
			"</span></div>" +
			'<div class="portal-gantt-popup-row"><span class="portal-gantt-popup-label">' +
			gantt_label("progress", "Progress") +
			'</span><span class="portal-gantt-popup-value">' +
			frappe.utils.escape_html(String(progress)) +
			"</span></div>" +
			'<div class="portal-gantt-popup-row"><span class="portal-gantt-popup-label">' +
			gantt_label("assignment", "Assignment") +
			'</span><span class="portal-gantt-popup-value">' +
			build_assignee_avatars(meta.assignees) +
			"</span></div></div>" +
			'<div class="portal-gantt-popup-actions">' +
			'<a href="' +
			task_url +
			'">' +
			gantt_label("open_task", "Open task") +
			" →</a>" +
			"</div></div>"
		);
	}

	function render_gantt() {
		if (project_gantt) {
			return;
		}

		if (!gantt_tasks.length) {
			$("#project-gantt").html(
				'<p class="text-muted text-center py-5">' +
					(gantt_label("no_tasks_gantt", "No tasks available for Gantt view")) +
					"</p>"
			);
			return;
		}

		load_gantt_lib(function () {
			project_gantt = new Gantt("#project-gantt", gantt_tasks, {
				bar_height: 32,
				bar_corner_radius: 4,
				view_mode: "Week",
				on_click: function () {
					/* popup handled by custom_popup_html */
				},
				custom_popup_html: function (task) {
					setTimeout(style_gantt_popup, 0);
					return build_gantt_popup(task);
				},
			});
			style_gantt_popup();
			apply_gantt_colors(gantt_tasks);
		});
	}

	function apply_gantt_colors(tasks) {
		var styles = tasks
			.filter(function (t) {
				return t.custom_class && t.color;
			})
			.map(function (t) {
				var class_name = t.custom_class;
				var bar_color = t.color;
				var progress_color = darken_color(bar_color, 0.25);
				return (
					".gantt .bar-wrapper." +
					class_name +
					" .bar { fill: " +
					bar_color +
					"; }" +
					".gantt .bar-wrapper." +
					class_name +
					" .bar-progress { fill: " +
					progress_color +
					"; }"
				);
			})
			.join("");

		if (styles) {
			$("#project-gantt-colors").remove();
			$('<style id="project-gantt-colors">' + styles + "</style>").prependTo("#tasks-gantt-view");
		}
	}

	function format_gantt_date(value) {
		if (!value) {
			return "";
		}

		var date_str;
		if (typeof value === "string") {
			date_str = value.split(" ")[0];
		} else {
			try {
				date_str = value.toISOString().slice(0, 10);
			} catch (e) {
				return String(value);
			}
		}

		if (frappe.datetime && frappe.datetime.str_to_user) {
			return frappe.datetime.str_to_user(date_str, false, true);
		}

		return date_str;
	}

	function darken_color(hex, amount) {
		if (!hex || hex.charAt(0) !== "#") {
			return "#0271a8";
		}
		var num = parseInt(hex.slice(1), 16);
		var r = Math.max(0, ((num >> 16) & 0xff) * (1 - amount));
		var g = Math.max(0, ((num >> 8) & 0xff) * (1 - amount));
		var b = Math.max(0, (num & 0xff) * (1 - amount));
		return (
			"#" +
			((1 << 24) + (Math.round(r) << 16) + (Math.round(g) << 8) + Math.round(b))
				.toString(16)
				.slice(1)
		);
	}

	function load_gantt_lib(callback) {
		if (window.Gantt) {
			callback();
			return;
		}

		var script = document.createElement("script");
		script.src = "/assets/frappe/node_modules/frappe-gantt/dist/frappe-gantt.min.js";
		script.onload = callback;
		document.head.appendChild(script);
	}

	$(".file-size").each(function () {
		$(this).text(frappe.form.formatters.FileSize($(this).text()));
	});
});

File: logicposintegration/templates/pages/projects.py
from __future__ import annotations

import json
from typing import Any

import frappe
from frappe.utils import add_days, formatdate, getdate, today

from erpnext.templates.pages.projects import get_attachments, get_timesheets

from logicposintegration.utils.portal_jinja import PRIORITY_COLORS, STATUS_COLORS

DEFAULT_GANTT_COLOR = "#02A8E5"

GANTT_STATUS_OPTIONS = [
	"Open",
	"Working",
	"Pending Review",
	"Overdue",
	"Completed",
	"Cancelled",
	"Template",
]

GANTT_PRIORITY_OPTIONS = ["Low", "Medium", "High", "Urgent"]


def get_gantt_i18n() -> dict[str, Any]:
	return {
		"status": frappe._("Status"),
		"priority": frappe._("Priority"),
		"progress": frappe._("Progress"),
		"assignment": frappe._("Assignment"),
		"open_task": frappe._("Open task"),
		"no_tasks_gantt": frappe._("No tasks available for Gantt view"),
		"status_labels": {option: frappe._(option) for option in GANTT_STATUS_OPTIONS},
		"priority_labels": {option: frappe._(option) for option in GANTT_PRIORITY_OPTIONS},
	}


def portal_user_can_access_project(project: str, user: str | None = None) -> bool:
	"""Portal: Project User no projeto, ou Portal User do Customer do projeto."""
	user = user or frappe.session.user
	if not project or not user or user == "Guest":
		return False
	if user == "Administrator":
		return True

	if frappe.db.exists("Project User", {"parent": project, "user": user}):
		return True

	customer = frappe.db.get_value("Project", project, "customer")
	if not customer:
		return False

	return bool(
		frappe.db.exists(
			"Portal User",
			{"parent": customer, "parenttype": "Customer", "user": user},
		)
	)


def get_context(context: Any) -> Any:
	project_name = frappe.form_dict.project
	user = frappe.session.user

	if user == "Guest" or not portal_user_can_access_project(project_name, user):
		raise frappe.PermissionError

	project_user = frappe.db.get_value(
		"Project User",
		{"parent": project_name, "user": user},
		["user", "view_attachments", "hide_timesheets"],
		as_dict=True,
	)

	context.no_cache = 1
	context.show_sidebar = True
	project = frappe.get_doc("Project", project_name)

	project.tasks = get_portal_tasks(
		project.name,
		search=frappe.form_dict.get("search"),
	)

	if project_user and not project_user.hide_timesheets:
		project.timesheets = get_timesheets(project.name, start=0, search=frappe.form_dict.get("search"))

	if project_user and project_user.view_attachments:
		project.attachments = get_attachments(project.name)

	context.doc = project
	context.status_colors = STATUS_COLORS
	context.priority_colors = PRIORITY_COLORS
	context.gantt_tasks = json.dumps(get_gantt_tasks(project.name))
	context.gantt_i18n = json.dumps(get_gantt_i18n(), default=str)
	context.kanban_columns = get_kanban_columns(project.name)
	return context


def get_portal_tasks(project, start=0, search=None, item_status=None):
	filters = {"project": project}
	if search:
		filters["subject"] = ("like", f"%{search}%")

	tasks = frappe.get_all(
		"Task",
		filters=filters,
		fields=[
			"name",
			"subject",
			"status",
			"priority",
			"modified",
			"_assign",
			"exp_end_date",
			"is_group",
			"parent_task",
		],
		limit_start=start,
		limit_page_length=100,
	)

	for task in tasks:
		if task.is_group:
			child_tasks = list(filter(lambda x: x.parent_task == task.name, tasks))
			if child_tasks:
				task.children = child_tasks

	return list(filter(lambda x: not x.parent_task, tasks))


def get_kanban_columns(project):
	statuses = [
		"Open",
		"Working",
		"Pending Review",
		"Overdue",
		"Completed",
	]

	tasks = frappe.get_all(
		"Task",
		filters={"project": project, "status": ("in", statuses)},
		fields=["name", "subject", "status", "exp_end_date", "progress", "priority", "color"],
		order_by="modified desc",
		limit=500,
	)

	columns = []
	for status in statuses:
		columns.append(
			{
				"status": status,
				"label": frappe._(status),
				"tasks": [task for task in tasks if task.status == status],
			}
		)

	return columns


def get_task_assignees(task):
	assignees = []
	if not task.get("_assign"):
		return assignees

	for user in json.loads(task._assign):
		details = frappe.db.get_value("User", user, ["full_name", "user_image"], as_dict=True)
		full_name = (details.full_name if details else None) or user
		assignees.append(
			{
				"user": user,
				"full_name": full_name,
				"user_image": details.user_image if details else None,
				"abbr": frappe.utils.get_abbr(full_name),
			}
		)

	return assignees


def get_gantt_color(color):
	if color and isinstance(color, str) and color.startswith("#") and len(color) >= 4:
		return color
	return DEFAULT_GANTT_COLOR


def get_gantt_tasks(project):
	project_doc = frappe.get_doc("Project", project)
	default_start = project_doc.expected_start_date or project_doc.actual_start_date or today()
	default_end = project_doc.expected_end_date or project_doc.actual_end_date or add_days(default_start, 7)

	tasks = frappe.get_all(
		"Task",
		filters={"project": project, "status": ("!=", "Cancelled")},
		fields=[
			"name",
			"subject",
			"status",
			"priority",
			"color",
			"exp_start_date",
			"exp_end_date",
			"progress",
			"depends_on_tasks",
			"_assign",
		],
		order_by="exp_start_date asc, creation asc",
		limit=500,
	)

	gantt_tasks = []
	for task in tasks:
		start = task.exp_start_date or default_start
		end = task.exp_end_date or default_end
		if end < start:
			end = add_days(start, 1)

		color = get_gantt_color(task.color)
		assignees = get_task_assignees(task)

		gantt_tasks.append(
			{
				"id": task.name,
				"name": task.subject or task.name,
				"start": formatdate(getdate(start), "yyyy-mm-dd"),
				"end": formatdate(getdate(end), "yyyy-mm-dd"),
				"progress": task.progress or 0,
				"dependencies": task.depends_on_tasks or "",
				"color": color,
				"custom_class": f"color-{color.lstrip('#')}",
				"status": task.status,
				"priority": task.priority or "",
				"assignees": assignees,
			}
		)

	return gantt_tasks
